File: leetcode/cpp/p26-remove-duplicates-from-sorted-array.cpp
// LeetCode problem 26. Remove duplicates from sorted array.
#include <algorithm>
#include <iostream>
#include <sstream>
#include <vector>

class Solution {
public:
    int removeDuplicates(std::vector<int>& nums)
    {
        const auto it = std::unique(nums.begin(), nums.end());
        return static_cast<int>(std::distance(nums.begin(), it));
    }
};

static std::string toString(std::vector<int> nums)
{
    std::ostringstream oss;
    for (const auto num : nums) {
        oss << num << ',';
    }
    return oss.str();
}

int main()
{
    struct TestCase {
        std::vector<int> nums;
        int exp;
    };
    TestCase testCases[] = {
        { {}, 0 },
        { { 1 }, 1 },
        { { 1, 2 }, 2 },
        { { 1, 1, 2 }, 2 },
        { { 1, 2, 2 }, 2 },
        { { 2, 2, 3, 4 }, 3 },
        { { 0, 0, 1, 1, 1, 2, 2, 3, 3, 4 }, 5 }
    };

    Solution s;
    for (auto& tc : testCases) {
        const auto ans = s.removeDuplicates(tc.nums);
        if (tc.exp != ans) {
            std::cout << "FAIL. prices: (nums: " << toString(tc.nums) << ")"
                      << ", exp: " << tc.exp
                      << ", ans: " << ans << "\n";
        }
    }
    return 0;
}

File: leetcode/rust/p26-remove-duplicates-from-sorted-array.rs
// LeetCode 26. Remove Duplicates from Sorted Array.
use std::convert::TryFrom;
use std::vec::Vec;

struct Solution {}

impl Solution {
    pub fn remove_duplicates(nums: &mut Vec<i32>) -> i32 {
        let mut k = 0;
        for i in 1..nums.len() {
            if nums[i] != nums[k] {
                k += 1;
                nums[k] = nums[i];
            }
        }
        return i32::try_from(k + 1).expect("Bad usize -> i32 conversion");
    }
}

fn main() {
    struct TestCase {
        nums: Vec<i32>,
        exp: Vec<i32>,
    }

    let mut test_cases = Vec::<TestCase>::new();
    test_cases.push(TestCase {
        nums: vec![1, 1, 2],
        exp: vec![1, 2],
    });
    test_cases.push(TestCase {
        nums: vec![0, 0, 1, 1, 1, 2, 2, 3, 3, 4],
        exp: vec![0, 1, 2, 3, 4],
    });

    for tc in test_cases.iter() {
        let mut act = tc.nums.clone();
        let k = Solution::remove_duplicates(&mut act);
        if tc.exp != act[..usize::try_from(k).expect("Bad i32->usize conversion")] {
            println!(
                "FAIL. nums: {:?}, exp: {:?}, act: {:?}, k: {}",
                tc.nums, tc.exp, act, k
            );
        }
    }
}
